src/circuit/circuit_zsa/gadget.rs

```diff
@@ -0,0 +1,227 @@
+//! Gadgets used in the Orchard circuit.
```

Comment:
From @PaulLaux:
lets unite this file with circuit/circuit_vanilla/gadget.rs and put the combined content into circuit/gadget.rs.
we can have 2 diffrent copies of derive_nullifier() (different names)

Reply:
We have additionally considered the idea of unifying the `gadget.rs` files (we briefly discussed it in the last meeting), and here are my findings and rationale for keeping them separate at this time:

- Upon closer inspection, the `gadget.rs` files for vanilla and ZSA variations are significantly different. They only share a small `assign_free_advice` function.
- The implementations of `super::Config` are programmatically distinct for each variation, as the `Config` struct is sourced from different modules (`circuit_vanilla.rs` and `circuit_zsa.rs`):
   - Initially, it seemed feasible to make the vanilla `Config` a base field in the ZSA `Config` with an additional `cond_swap_config` field. However, further analysis revealed substantial differences in their fields.
   - While the fields have the same names and types, most of these types differ in definition between the vanilla and ZSA versions (`NoteCommitConfig`, `EccConfig`, `MerkleConfig`, `SinsemillaConfig`).
   - These differences arise because the types will be generics with different arguments once we switch to the backward-compatible `halo2` crate, making them almost entirely distinct (except `NoteCommitConfig`, which is defined in the `orchard` crate, but defined differently for vanilla and zsa versions of the circuit).

Attempting to unify the `Config` structs now by making them generic and propagating these generics throughout (`EccConfig`, `MerkleConfig`, `SinsemillaConfig`) seems overly complex at this stage (it also requires converting `NoteCommitConfig` to be a generic).

As discussed, we can revisit the possibility of generalizing the `Config` struct and unifying the `gadget.rs` files after transitioning to the backward-compatible version of `halo2`.